aten/src/ATen/native/nested/cuda/NestedTensorTransformerFunctions.cpp

```diff
@@ -218,36 +211,201 @@ Tensor NestedTensor_to_padded_tensor_cuda(
+      {batch_size + 1}, TensorOptions().device(at::kCPU).dtype(at::kInt));
+
+  auto* sizes_ptr = sizes.data_ptr<int64_t>();
+  auto* cumulative_seqlen_ptr = cumulative_seqlen.data_ptr<int32_t>();
+
+  int32_t sum = 0;
+  int64_t max_seqlen = -1;
+  cumulative_seqlen_ptr[0] = sum;
+  for (const auto i : c10::irange(batch_size)) {
+    // Calculate the cumulative sum of the sequence lengths
+    auto current_seq_len = sizes_ptr[(i * size_tensor_stride) + seq_len_dim];
+    sum += current_seq_len;
+    cumulative_seqlen_ptr[i + 1] = sum;
+
+    // Find the max element while we traverse
+    max_seqlen = std::max(max_seqlen, current_seq_len);
+  }
+  // Send to GPU, this is pretty light weight calc for normal batch size
+  // but maybe this needs to be on gpu
+  cumulative_seqlen = cumulative_seqlen.to(TensorOptions().device(at::kCUDA));
+  return std::tuple<Tensor, int64_t>{cumulative_seqlen, max_seqlen};
+}
+bool is_chunked_nested_tensor_group(const Tensor& query, const Tensor& key, const Tensor& value){
+  return (query.storage().is_alias_of(key.storage())) && (query.storage().is_alias_of(value.storage()));
+}
+
+std::tuple<Tensor, Tensor> flash_attention_helper_nested_unpacked(
     const Tensor& query,
     const Tensor& key,
     const Tensor& value,
-    const Tensor& cumulative_sequence_length_q,
-    const Tensor& cumulative_sequence_length_k,
-    const int64_t max_seqlen_batch_q,
-    const int64_t max_seqlen_batch_k,
     double dropout_p,
-    bool causal) {
-#if defined(USE_FLASH_ATTENTION)
-  auto softmax_scale = std::pow(query.size(-1), -0.5);
-  std::vector<Tensor> output = fmha::mha_fwd(
-      query,
-      key,
-      value,
+    bool need_atten_weights,
+    bool is_causal) {
+  // Query (Batch x Num_heads x {Q_seq_len}  x Dim_per_head)
+  // Key   (Batch x Num_heads x {KV_seq_len} x Dim_per_head)
+  // Value (Batch x Num_heads x {KV_seq_len} x Dim_per_head)
+  const int64_t num_heads = query.size(1);
+  const int64_t head_dim = query.size(3);
+
+  Tensor q_t = query.transpose(1, 2);
+  Tensor k_t = key.transpose(1, 2);
+  Tensor v_t = value.transpose(1, 2);
+
+  auto cumulative_and_max_q = cumulative_and_max_seq_len(q_t, 0);
+  auto cumulative_and_max_k = cumulative_and_max_seq_len(k_t, 0);
+
+  // K and V have to have the same Nnz, should probably torch_check
+  // assume in order to not iterate over v
+
+  Tensor cumulative_sequence_length_q = std::get<0>(cumulative_and_max_q);
+  Tensor cumulative_sequence_length_k = std::get<0>(cumulative_and_max_k);
+
+  const int64_t max_seqlen_batch_q = std::get<1>(cumulative_and_max_q);
+  const int64_t max_seqlen_batch_k = std::get<1>(cumulative_and_max_k);
+
+  const int64_t Nnz_q  = cumulative_sequence_length_q[-1].item<int64_t>();
+  const int64_t Nnz_kv = cumulative_sequence_length_k[-1].item<int64_t>();
+
+  Tensor query_buffer_reshaped;
+  Tensor key_buffer_reshaped;
+  Tensor value_buffer_reshaped;
+
+  if (is_chunked_nested_tensor_group(query, key, value)) {
```

Comment:
I don't think this is enough of a condition to ensure that the `as_strided` calls yield correct output.

We don't need to check that query, key and value Storage are all aliases of each other, we just need to check whether that call to `as_strided` is allowed.

Reply:
That means making sure the strides `{3 * num_heads * head_dim, head_dim, 1}` and offset `x * packed_offset` yield a valid view of the data.